File: ios/BiblosSurfer/Services/ServiceProvider.swift
//
//  ServiceProvider.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 22/08/2026.
//

import Foundation

protocol ServiceProviderProtocol: AnyObject {
    var bookService: BookStoreProtocol { get }
    var libraryService: LibraryServiceProtocol { get }
    var publicationOpener: PublicationOpeningServiceProtocol { get }
    var settings: ReaderSettingsStore { get }
}

final class ServiceProvider: ServiceProviderProtocol {
    let bookService: BookStoreProtocol
    let libraryService: LibraryServiceProtocol
    let publicationOpener: PublicationOpeningServiceProtocol
    let settings: ReaderSettingsStore

    init() {
        publicationOpener = PublicationOpeningService()
        bookService = SwiftDataBookStore()
        settings = ReaderSettingsStore()
        libraryService = LibraryService(
            opener: publicationOpener,
            bookStore: bookService
        )
    }
}

File: android/app/src/main/java/miko/biblossurfer/data/ServiceProvider.kt
package miko.biblossurfer.data

import android.content.Context

interface ServiceProviderProtocol {
    val bookService: BookStoreProtocol
    val libraryService: LibraryServiceProtocol
    val publicationOpener: PublicationOpeningServiceProtocol
    val settings: ReaderSettingsStore
}

class ServiceProvider(context: Context) : ServiceProviderProtocol {
    override val publicationOpener: PublicationOpeningServiceProtocol =
        PublicationOpeningService(context)
    override val bookService: BookStoreProtocol = RoomBookStore(context)
    override val settings: ReaderSettingsStore =
        ReaderSettingsStore(context.getSharedPreferences("miko.biblossurfer", Context.MODE_PRIVATE))
    override val libraryService: LibraryServiceProtocol = LibraryService(
        context = context,
        opener = publicationOpener,
        bookStore = bookService,
    )
}

class MockServiceProvider(context: Context) : ServiceProviderProtocol {
    override val publicationOpener: PublicationOpeningServiceProtocol =
        PublicationOpeningService(context)
    override val bookService: BookStoreProtocol = LocalBookService()
    override val settings: ReaderSettingsStore =
        ReaderSettingsStore(
            context.getSharedPreferences("miko.biblossurfer.mock", Context.MODE_PRIVATE)
        )
    override val libraryService: LibraryServiceProtocol = UITestStubLibraryService(context)
}
